### ISSUE TITLE
Disulfide bonds are broken in CHARMM36

### ISSUE THREAD

**Issue by peastman** [pandegroup/openmm#2039]
It appears we have another critical force field error: disulfide bonds don't work in CHARMM36 either.  They're supposed to be handled by the DISU patch, but that patch somehow was omitted from the converted force field.

**Comment by jchodera** [comment]
Looking into this now.

**Comment by jchodera** [comment]
@peastman: It looks like it is likely a logic error in [`_determine_valid_patch_combinations()`](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L547-L587) in ParmEd. I don't think this logic considered the possibility of patches that must be applied to two residues. 

Can you remind me how OpenMM handles that case? Would both residues flag that with `AllowPatch`? And any suggestions on a scalable way to automatically test for patches that are compatible with two residues? We can't simply test combinations of patches that are compatible with single residues because a patch that requires two residues will fail if applied to a single residue using our current criteria, applied [here](https://github.com/ParmEd/ParmEd/blob/b393f747f9bf76af2ec5fe243f3f41ecdd9ac06a/parmed/modeller/residue.py#L425-L513).

**Comment by peastman** [comment]
The syntax is described at http://docs.openmm.org/latest/userguide/application.html#patches.

Patches that apply to multiple residues are indicated in the rtf file by prefixing each atom name with a number.  For example, the DISU patch specifies

```
GROUP                  
ATOM 1CB  CT2    -0.10 !
ATOM 1SG  SM     -0.08 !           2SG--2CB--
GROUP                  !          /
ATOM 2SG  SM     -0.08 ! -1CB--1SG
ATOM 2CB  CT2    -0.10 !
DELETE ATOM 1HG1   
DELETE ATOM 2HG1   
BOND 1SG 2SG   
```

So we can see there are two residues, and each of them has atoms named CB, SG, and HG1.

**Comment by jchodera** [comment]
We definitely overlooked the multiple-residue-patch case in engineering the code that attempts to automate identification of compatible patches. Any ideas on a workable algorithm? The code already takes many minutes just to check npatches x nresidues?

In the short term, the only path forward I can easily see is to manually whitelist patches that apply to multiple residues, and even that looks like it will take a while to implement and test if ParmEd doesn't currently understand CHARMM patches that apply to atoms from multiple residues.

What about a quick ffxml manual addition to fix the disulfide issue while we discuss how to automate the process in a more robust way?

**Comment by peastman** [comment]
> What about a quick ffxml manual addition to fix the disulfide issue while we discuss how to automate the process in a more robust way?

Ok with me.

> The code already takes many minutes just to check npatches x nresidues?

This shouldn't affect that.  Applying multi-residue patches is the very last thing it tries after first trying standard templates, then single residue patches.  So any topology that currently works will not get any slower.

**Comment by jchodera** [comment]
@peastman: It looks like it is likely a logic error in [`_determine_valid_patch_combinations()`](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L547-L587) in ParmEd. I don't think this logic considered the possibility of patches that must be applied to two residues. 

Can you remind me how OpenMM handles that case? Would both residues flag that with `AllowPatch`? And any suggestions on a scalable way to automatically test for patches that are compatible with two residues? We can't simply test combinations of patches that are compatible with single residues because a patch that requires two residues will fail if applied to a single residue using our current criteria, applied [here](https://github.com/ParmEd/ParmEd/blob/b393f747f9bf76af2ec5fe243f3f41ecdd9ac06a/parmed/modeller/residue.py#L425-L513).

**Comment by jchodera** [comment]
> This shouldn't affect that. Applying multi-residue patches is the very last thing it tries after first trying standard templates, then single residue patches. So any topology that currently works will not get any slower.

Oh, I was referring to the parameter conversion code. See [`_determine_valid_patch_combinations()`](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L547-L587).

**Comment by jchodera** [comment]
There aren't many two-residue patches *except* for [the carbohydrates](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/top_all36_carb.rtf), where there seem to the perhaps a few dozen. 

In most cases, the patch deletes some atoms from each residue, and then adds atoms and/or bonds. This could be relatively quick to evaluate by quickly establishing which set of residues contains the atoms being deleted before validating whether the rest of the patch can be applied.

The problem with this idea is that some patches, like [these](https://github.com/choderalab/openmm-forcefields/blob/master/charmm/toppar/top_all36_prot.rtf#L1733-L1774), don't delete any atoms, and this scheme would require evaluating all possible combinations of residues. We can add a second level of checking to make sure the residues contain atoms in the `BOND` directives, but this starts to get rather complicated. But those two levels might make it fast enough to evaluate all remaining combinations to make sure this results in a valid residue after patching.

This would still take a decent amount of time to implement and debug.

**Comment by peastman** [comment]
How about this then.  For the moment, I'll manually add a patch to handle disulfides.  That should be good enough for a patch release.  Then we can implement the changes in ParmEd (and ForceField if necessary) to handle all of the two residue templates, and that will be included in 7.3.

> Also, if you can point me to the example that failed, I can add that to our test suite to make sure we can test the fix.

It was a user who found it, not in one of our examples.  You should be able to use the same test case you're using for Amber.

**Comment by jchodera** [comment]
Sounds good.

We've got the CUDA 9.1 build framework set up for the dev builds now, so we should be able to release 7.3 for CUDA 9.1 (and possibly other cudas if we get th features thing to work).
